src/main/java/org/terasology/climateConditions/igloo/IglooProvider.java

```diff
@@ -0,0 +1,102 @@
+import org.terasology.math.geom.Rect2i;
+import org.terasology.math.geom.Vector2i;
+import org.terasology.utilities.procedural.Noise;
+import org.terasology.utilities.procedural.WhiteNoise;
+import org.terasology.world.generation.Border3D;
+import org.terasology.world.generation.Facet;
+import org.terasology.world.generation.FacetBorder;
+import org.terasology.world.generation.FacetProviderPlugin;
+import org.terasology.world.generation.GeneratingRegion;
+import org.terasology.world.generation.Produces;
+import org.terasology.world.generation.Requires;
+import org.terasology.world.generation.facets.SeaLevelFacet;
+import org.terasology.world.generation.facets.SurfaceHeightFacet;
+import org.terasology.world.generation.facets.base.BaseFieldFacet2D;
+import org.terasology.world.generator.plugin.RegisterPlugin;
+
+@RegisterPlugin
+@Requires({
+        @Facet(value = SurfaceHeightFacet.class, border = @FacetBorder(sides = Igloo.SIZE * 2)),
+        @Facet(value = BiomeFacet.class, border = @FacetBorder(sides = Igloo.SIZE * 2)),
+        @Facet(value = SeaLevelFacet.class, border = @FacetBorder(sides = Igloo.SIZE * 2))
+})
+@Produces(IglooFacet.class)
+public class IglooProvider implements FacetProviderPlugin {
+    private Noise noise;
+
+    @Override
+    public void setSeed(long seed) {
+        noise = new WhiteNoise(seed + 3);
+    }
+
+    /**
+     * Places the Igloo Structure in the Snow Biome with a very low probability of spawn.
+     */
+    @Override
+    public void process(GeneratingRegion region) {
+
+        Border3D border = region.getBorderForFacet(IglooFacet.class).extendBy(0, Igloo.SIZE * 2,
+                Igloo.SIZE * 2);
+
+        IglooFacet facet = new IglooFacet(region.getRegion(), border);
+        SurfaceHeightFacet surfaceHeightFacet = region.getRegionFacet(SurfaceHeightFacet.class);
+        SeaLevelFacet seaLevelFacet = region.getRegionFacet(SeaLevelFacet.class);
+        BiomeFacet biomeFacet = region.getRegionFacet(BiomeFacet.class);
+
+        Rect2i worldRegion = surfaceHeightFacet.getWorldRegion();
+
+        for (int wz = worldRegion.minY(); wz <= worldRegion.maxY(); wz++) {
+            for (int wx = worldRegion.minX(); wx <= worldRegion.maxX(); wx++) {
+                int surfaceHeight = TeraMath.floorToInt(surfaceHeightFacet.getWorld(wx, wz));
+                int seaLevel = seaLevelFacet.getSeaLevel();
+                // check if height is within this region
+                if (surfaceHeight >= facet.getWorldRegion().minY() &&
+                        surfaceHeight <= facet.getWorldRegion().maxY()) {
+                    // Sea Level + 96 is the height at which snow blocks are placed in all biomes.
+                    if (noise.noise(wx, wz) > 0.9999f && (surfaceHeight >= seaLevel + 96f ||
+                            biomeFacet.getWorld(wx, wz).getId().equals(CoreBiome.SNOW))) {
+                        int lowestY = getLowestY(surfaceHeightFacet, new Vector2i(wx, wz),
```

Comment:
placing the igloo "in the ground" definitely works fine here. In the future, we might improve placement to flat areas and don't place them into slopes.

Reply:
Yeah sure but from what I've observed the way the world is generated right now (as far as perlin generator is concerned), there aren't a lot of flat areas, there are just constantly rising mountains everywhere no flat areas with stable heights. Thus very low probability of igloo facet placing and then the even lower possibility that the found place would be a flat area would probably end up resulting in close to no igloos(it's actually still 1-2 in one snow biome area but then it would be further lower)... I think having well balanced flat areas at various heights should be a task with higher priority if and when possible.